### PR TITLE
feat(metricprovider): add prometheus range query support

### PR DESCRIPTION
**Pull request by mclarke47**
fixes: https://github.com/argoproj/argo-rollouts/issues/3702

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/) with a list of types and scopes found [here](https://github.com/argoproj/argo-rollouts/blob/master/.github/workflows/pr-title-check.yml), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

**Comment by mclarke47** [comment]
It would be nice if we could parameterize `lookBackDuration` through args also, but not 100% sure how to do this

**Comment by mclarke47** [comment]
Query ranges are syntactic sugar for the query endpoint as it says in that blog post. However I think there are 2 main benefits to adding query ranges: 

* They allow writing queries similar to how they are viewed in graphs and so IMO makes it more intuitive, and you can re-use queries from other sources without needing to first convert them to queries that would work with the query endpoint
* Query ranges support the step argument which allows you to specify the windows in which the queries over the range are executed so you get a little more control over how the data it returned from prometheus

**Comment by zachaller** [comment]
That's fair, I do see it being a bit easier to reason about, also do you think it make sense to expose the start and end parameters vs hiding that behind `lookBackDuration`?

**Comment by mclarke47** [comment]
I think that's would be good, although it should probably be relative to `time.Now()` otherwise we would be hard coding a datetime which seems not very useful.

@zachaller is there a way to access args of the analysis template inside the provider so that we can parameterize things?

**Comment by zachaller** [comment]
> I think that's would be good, although it should probably be relative to `time.Now()` otherwise we would be hard coding a datetime which seems not very useful.
> 
> @zachaller is there a way to access args of the analysis template inside the provider so that we can parameterize things?

Pretty sure this works today because we do things like:
```
provider:
        prometheus:
          address: "http://prometheus:{{args.prometheus-port}}"
          query: |-
            (quantile(0.5, quantile_over_time(0.5, namespace_pod_cpu_utilization{namespace="{{args.namespace}}", pod=~".*-{{args.canary-hash}}-.*"}[11m])))
```

Might be worth checking if it still works with your code as well just incase the template system explicity looks for those fields etc.

**Review by IvanManz — APPROVED** [review]

**Comment by mclarke47** [comment]
Will double check if this already works

**Comment by mclarke47** [comment]
Confirming:

```
...
  - name: lookback_duration
    value: "5m"
  - name: lookback_step
    value: "1m"
...    
    provider:
      prometheus:
        rangeQuery:
          lookBackDuration: "{{args.lookback_duration}}"
          step: "{{args.lookback_step}}"
...          
```
does work

**Comment by mclarke47** [comment]
There are time functions in [expr](https://expr-lang.org/docs/language-definition#date-functions), I'm going to change the API to be start/end/step, with start's default being `now()`. I'll test this out then update the docs.

**Comment by mclarke47** [comment]
hmmm, it seems this expression logic is only evaluated in success/failure conditions not in the actual fields of the provider, only args substitution works there.

I could:

* Stick with a lookback duration as is currently implemented
* Put in some logic to Evaluate these fields with `util/evaluate`

WDYT @zachaller?


Screenshot 2024-07-08 at 2 15 10 PM

**Comment by mclarke47** [comment]
Hey @zachaller, I pushed that change to treat the start/end fields as expr expressions. 🙏

File: utils/evaluate/evaluate.go [hunk 1]
```diff
@@ -56,6 +57,41 @@ func EvaluateResult(result any, metric v1alpha1.Metric, logCtx logrus.Entry) (v1
 	return v1alpha1.AnalysisPhaseSuccessful, nil
 }
 
+func EvalTime(expression string) (time.Time, error) {
+	var err error
+
+	env := map[string]any{
+		"asInt":   asInt,
+		"asFloat": asFloat,
```
Comment (zachaller) [inline utils/evaluate/evaluate.go:65]: Let's remove asInt, asFloat and just use the built in funcs https://expr-lang.org/docs/language-definition#int

File: utils/evaluate/evaluate.go [hunk 2]
```diff
@@ -56,6 +57,41 @@ func EvaluateResult(result any, metric v1alpha1.Metric, logCtx logrus.Entry) (v1
...
+	output, err := expr.Run(program, env)
+	if err != nil {
+		return time.Time{}, unwrapFileErr(err)
+	}
+
+	switch val := output.(type) {
+	case time.Time:
+		return val, nil
+	default:
+		return time.Time{}, fmt.Errorf("expected string, but got %T", val)
```
Comment (zachaller) [inline utils/evaluate/evaluate.go:91]: expected time.Time

File: metricproviders/prometheus/prometheus.go [hunk 3]
```diff
@@ -112,29 +136,51 @@ func (p *Provider) GarbageCollect(run *v1alpha1.AnalysisRun, metric v1alpha1.Met
...
 
+func sampleValuesToFloatSlice(SampleValues []model.SampleValue) []float64 {
+	results := make([]float64, 0, len(SampleValues))
+	for _, s := range SampleValues {
+		results = append(results, float64(s))
+	}
+	return results
+}
+
+func sampleValuesToResultStr(SampleValues []model.SampleValue) string {
```
Comment (zachaller) [inline metricproviders/prometheus/prometheus.go:147]: Can we lowercase SampleValues

**Comment by zachaller** [comment]
Just a few small changes then this LGTM

File: metricproviders/prometheus/prometheus.go (hunk 3 above)
Comment (mclarke47) [reply metricproviders/prometheus/prometheus.go]: fixed!

File: metricproviders/prometheus/prometheus.go [hunk 4]
```diff
@@ -112,29 +136,51 @@ func (p *Provider) GarbageCollect(run *v1alpha1.AnalysisRun, metric v1alpha1.Met
 	return nil
 }
 
+func sampleValuesToFloatSlice(SampleValues []model.SampleValue) []float64 {
```
[the remark this answers is not in the document]
Comment (mclarke47) [reply metricproviders/prometheus/prometheus.go]: fixed!

File: utils/evaluate/evaluate.go (hunk 2 above)
Comment (mclarke47) [reply utils/evaluate/evaluate.go]: fixed!

File: utils/evaluate/evaluate.go (hunk 1 above)
Comment (mclarke47) [reply utils/evaluate/evaluate.go]: fixed!

**Comment by mclarke47** [comment]
@zachaller have address your comments 🙏

**Comment by zachaller** [comment]
Thank you for the contribution!

**Pull request merged**